src/log.cpp

```diff
@@ -229,45 +227,50 @@ const std::string Logger::getLevelLabel(LogLevel lev)
 		"VERBOSE",
 		"TRACE",
 	};
-	assert(lev < LL_MAX && lev >= 0);
 	static_assert(ARRLEN(names) == LL_MAX,
 		"mismatch between loglevel names and enum");
+	assert(lev < LL_MAX && lev >= 0);
 	return names[lev];
 }
 
 LogColor Logger::color_mode = LOG_COLOR_AUTO;
 
-const std::string Logger::getThreadName()
+const std::string &Logger::getThreadName()
 {
-	std::map<std::thread::id, std::string>::const_iterator it;
-
 	std::thread::id id = std::this_thread::get_id();
-	it = m_thread_names.find(id);
+
+	auto it = m_thread_names.find(id);
 	if (it != m_thread_names.end())
 		return it->second;
 
-	std::ostringstream os;
-	os << "#0x" << std::hex << id;
-	return os.str();
+	thread_local std::string fallback_name;
+	if (!fallback_name.empty()) {
+		std::ostringstream os;
+		os << "#0x" << std::hex << id;
+		fallback_name = os.str();
+	}
+	return fallback_name;
 }
 
 void Logger::log(LogLevel lev, const std::string &text)
 {
-	if (m_silenced_levels[lev])
+	if (isLevelSilenced(lev))
 		return;
 
-	const std::string thread_name = getThreadName();
-	const std::string label = getLevelLabel(lev);
+	const std::string &thread_name = getThreadName();
+	const char *label = getLevelLabel(lev);
 	const std::string timestamp = getTimestamp();
-	std::ostringstream os(std::ios_base::binary);
-	os << timestamp << ": " << label << "[" << thread_name << "]: " << text;
 
-	logToOutputs(lev, os.str(), timestamp, thread_name, text);
+	std::string line = timestamp;
+	line.append(": ").append(label).append("[").append(thread_name)
+		.append("]: ").append(text);
```

Comment:
At this point, `snprintf` might be easier to use..?

Reply:
true, but that's not C++-y :smiley:

Reply:
When you're given a hammer, everything is a nail. Sometimes problems have easier solutions than a hammer but I'll leave this one up to you.